dali/operators/generic/reduce/histogram.h

```diff
@@ -0,0 +1,248 @@
+  auto raxis = reduction_axes.rbegin();
+
+  for (; raxis != reduction_axes.rend(); ++raxis, --dim) {
+    if (dim != *raxis)
+      break;
+    // dimensions_to_collapse.insert(dimensions_to_collapse.begin(), dim);
+  }
+
+  std::vector<int> axes_to_transpose;
+
+  // All axes that couldn't be collapsed, need to be transposed so we can collapse those together.
+  int naxes_to_transpose = std::distance(raxis, reduction_axes.rend());
+
+  axes_to_transpose.resize(naxes_to_transpose);
+
+  auto raxes_start = reduction_axes.begin() + naxes_to_transpose;
+  std::copy(reduction_axes.begin(), raxes_start, axes_to_transpose.begin());
+
+  return axes_to_transpose;
+}
+
+class TransposeForReduction {
+
+};
+
+}  // namespace hist_detail
+}  // namespace dali
+
+
+
+namespace dali {
+
+template <typename Backend>
+class Histogram : public Operator<Backend>, detail::AxesHelper {
+ public:
+  explicit inline Histogram(const OpSpec &spec)
+      : Operator<Backend>(spec), detail::AxesHelper(spec) {}
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  ~Histogram() override = default;
+
+  void PrepareInputShapesForReduction(const TensorListShape<> &input_shapes, int hist_dim = 1) {
+    std::vector<TensorShape<>> reduction_input_shapes;
+    reduction_input_shapes.reserve(input_shapes.num_samples());
+
+    const int ndims = input_shapes.sample_dim();
+
+    // TODO: support higher dimentionality histograms (hist_dim)
+    std::array<std::pair<int, int>, 1> reduced_inner_dim(
+        {std::make_pair(ndims - axes_.size() - 1, ndims - 1)});
+
+    for (int i = 0; i < input_shapes.num_samples(); ++i) {
+      // Collapse inner dimensions
+      auto shape = collapse_dims(input_shapes.tensor_shape(i), reduced_inner_dim);
+      reduction_input_shapes.push_back(std::move(shape));
+    }
+
+    reduced_input_shapes_ = TensorListShape<>(reduction_input_shapes);
+  }
+
+  // For all subsequent dimensions go through the list and find axes not being
+  // reduced
+  SmallVector<int, 6> GetNonReductionAxes(const int ndims) const {
+    SmallVector<int, 6> non_reduction_axes;
+
+    int rax_search_start = 0;
+
+    for (int dim = 0; dim < ndims; ++dim) {
+      bool found = false;
+      int rax = rax_search_start;
+
+      while (rax < axes_.size() && axes_[rax] <= dim) {
+        // Since axes_ are sorted, we know we can skip ahead searching
+        // of next dimension to at least to next reduction axis
+        rax_search_start = rax + 1;
+
+        if (axes_[rax] == dim) {
+          found = true;
+          break;
+        }
+        ++rax;
+      }
+
+      if (!found) {
+        non_reduction_axes.push_back(dim);
+      }
+    }
+
+    return non_reduction_axes;
+  }
+
+  void PrepareInputShapesForTranspose(const TensorListShape<> &input_shapes, int hist_dim = 1) {
+    const int ndims = input_shapes.sample_dim();
+
+    auto shape_span
+    auto non_rediction_axes = GetNonReductionAxes(ndims);
```

Comment:
```suggestion
    auto non_reduction_axes = GetNonReductionAxes(ndims);
```
However, how about using bit masks for reduced axes? This would greatly simplify the code. You'd create a mask once and then you'd just check if given axis is marked as reduced or not. We even have a bitmask class, so there's no need to manually shift/mask stuff.